contract/models/contract.py

```diff
@@ -412,11 +413,28 @@ def _get_lines_to_invoice(self, date_ref):
         :return: contract lines (contract.line recordset)
         """
         self.ensure_one()
-        return self.contract_line_ids.filtered(
-            lambda l: not l.is_canceled
-            and l.recurring_next_date
-            and l.recurring_next_date <= date_ref
-        )
+
+        def can_be_invoiced(l):
+            return (not l.is_canceled and l.recurring_next_date
+                    and l.recurring_next_date <= date_ref)
+
+        res = current_section = self.env['contract.line']
+        add_current_section = add_next_note = False
+        for line in self.contract_line_ids.sorted('sequence'):
```

Comment:
Why do you need `sorted`?

Reply:
If I have contract lines sorted by 'sequence', iterating those lines with a for loop I can assure that lines that go after a certain section correspond to that section and lines inmediatly before a note correspond to that note. With that, i can determine with a single loop, which section/Notes and which product lines can be invoiced.